### ISSUE TITLE
Runtime exception when creating PKCS11::Module with Softhsm2 library under Windows

### ISSUE THREAD

**Issue by oliviervibe** [randombit/botan#2918]
I have used successfully the SoftHsm2 PKCS11 implementation as a raw PKCS library to generate AES Keys and encrypt/decrypt with them.

I'm trying to use Botan as a wrapper to test an issue I have with the Linux implementation when using the LowLevel API to generate AES keys - it always returns an AttributeValueInvalid.

The following code throws an exception on the Module creation
Unhandled exception at 0x00007FFDE9714F69 in TestPkcsBotan.exe: Microsoft C++ exception: std::bad_alloc at memory location 0x000000E50FEFF9E0.

int main()
{
    PKCS11::Module module("softhsm2 - x64.dll");
}

I have compiled the library using Python and nmake as described in the documentation.

Is there a precompiled version of Botan like the one that can be installed on Linux? 

Any working code generating AES key with the LowLevel API would also really help me as I have tried everything and I can't find any error with my parameters as I simply followed the PKCS specs, and the code works with direct calls to the PKCS lib. 

Thanks

**Comment by randombit** [comment]
> I have used successfully the SoftHsm2 PKCS11 implementation as a raw PKCS library to generate AES Keys and encrypt/decrypt with them.
> I'm trying to use Botan as a wrapper to test an issue I have with the Linux implementation when using the LowLevel API to generate AES keys - it always returns an AttributeValueInvalid.

Can you post your two versions (full source) - it's possible we are encoding an attribute incorrectly or something. If you are seeing this on Linux I should be able to replicate it.

> Unhandled exception at 0x00007FFDE9714F69 in TestPkcsBotan.exe: Microsoft C++ exception: std::bad_alloc at memory location 0x000000E50FEFF9E0.

This looks to me like a problem caused by the DLL and your binary using different MSVC runtimes. This is purely a guess, but it's hard to see why this simple call would otherwise cause a `bad_alloc`.

> Is there a precompiled version of Botan like the one that can be installed on Linux?

As far as I know there is no Windows binary distribution (but I don't use Windows so could not say for certain).

**Comment by oliviervibe** [comment]
Thanks, the most important is under Linux.

I'm going to extract the source I use to generate the AES under Linux and post the complete code.

As for Windows I just recompiled the Botan lib with the compiler of Visual Studio 2019 using make as stated and the program (one line!) is also compiled with Visual Studio 2019.  And I still get the same error.

**Comment by randombit** [comment]
> As for Windows I just recompiled the Botan lib with the compiler of Visual Studio 2019 using nmake as stated and the program (one line!) is also compiled with Visual Studio 2019. And I still get the same error.

Right, but are the runtimes the same? https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170

MSVC ships with 4 options (6 if you count DLLs), which are all completely incompatible, and if you mix them you typically just get crashes and memory corruption. I truly have no idea why they do this.

You can change which one Botan uses with `--msvc-runtime={MT,MD,MTd,MDd}` I think we default to `MT` but am not sure tbh.

**Comment by lieser** [comment]
Regarding precompiled Botan binaries on Windows:
The default repository of the Conan package manager has a package for Botan (https://conan.io/center/botan). And it does store some precompiled configurations.

There also exists a vcpkg package of Botan. But haven't used that so far, so don't know if they also store some precompiled configurations.

**Comment by oliviervibe** [comment]
I solved the problem!

While I was preparing the program I noticed something. 

In my Windows test the function I call takes the keySize parameter as a CK_ULONG while in the Linux code I used by mistake an uint32_t. 

As the parameter is passed by address... the value in read from the Attribute structure is then incorrect!

It now works.

I will check the solution you give to compile with Visual Studio. The most important for me was to have it working with Linux and it was a a silly mistake of mine!

Thanks for your help.
